measurement_kit.go

```diff
@@ -30,61 +33,86 @@ type NettestOptions struct {
 
 // Nettest is a wrapper for running a particular nettest
 type Nettest struct {
-	Name    string
-	Options NettestOptions
+	Name           string
+	Options        NettestOptions
+	DisabledEvents []string
 }
 
 type taskData struct {
-	EnabledEvents []string       `json:"enabled_events"`
-	Type          string         `json:"type"`
-	Verbosity     string         `json:"verbosity"`
-	Options       NettestOptions `json:"options"`
+	DisabledEvents []string       `json:"disabled_events"`
+	Type           string         `json:"type"`
+	Verbosity      string         `json:"verbosity"`
+	Options        NettestOptions `json:"options"`
 }
 
-var handleLock sync.Mutex
-var handleVals = make(map[int]func(interface{}))
-var handleIndex int
-
-func newHandle(v func(interface{})) int {
-	handleLock.Lock()
-	defer handleLock.Unlock()
-	i := handleIndex
-	handleIndex++
-	handleVals[i] = v
-	return i
+var handleMu sync.Mutex
+var handleMap = make(map[string][]interface{})
+
+// Event is an event fired from measurement_kit
+type Event struct {
+	Key   string      `json:"key"`
+	Value interface{} `json:"value"`
 }
 
-func notifyEventHandlers(event interface{}) {
-	for i := 0; i < handleIndex; i++ {
-		handleVals[i](event)
+func fire(s string, e Event) error {
+	parts := strings.Split(s, ".")
+	handles := make([]interface{}, 0)
+
+	handleMu.Lock()
+	// Add the literal match
+	hs, ok := handleMap[s]
+	if ok {
+		handles = append(handles, hs...)
 	}
+
+	// Look for wildcards such as foo.bar.*
+	for i := 1; i < len(parts); i++ {
+		hn := fmt.Sprintf("%s.*", strings.Join(parts[0:i], "."))
+		hs, ok := handleMap[hn]
+		if ok {
+			handles = append(handles, hs...)
+		}
+	}
+	handleMu.Unlock()
+
+	for _, handle := range handles {
+		f := reflect.ValueOf(handle)
+		args := make([]reflect.Value, 1)
+		args[0] = reflect.ValueOf(e)
+
+		// XXX should I do this call inside of a goroutine?
```

Comment:
I have no strong opinion on this. Perhaps we can see what martini or gin do? Without considering the application context, I may be worried about creating many coroutines. Maybe even an internal channel to post events on and a pool of coroutines to dispatch the callbacks can be good.